src/MIDebugEngine/Engine.Impl/Disassembly.cs

```diff
@@ -305,17 +305,153 @@ private void DeleteRangeFromCache(DisassemblyBlock block)
             if (results.ResultClass != ResultClass.done)
             {
                 return null;
             }
 
-            return DecodeDisassemblyInstructions(results.Find<ValueListValue>("asm_insns").AsArray<TupleValue>());
+            DisasmInstruction[] instructions =  DecodeDisassemblyInstructions(results.Find<ValueListValue>("asm_insns").AsArray<TupleValue>());
+
+            if (instructions != null && instructions.Length != 0)
+            {
+                // Map 'Symbol' (Function Name) to Address Range
+                Dictionary<string, DisasmAddressRange> funcToAddressRange = new Dictionary<string, DisasmAddressRange>();
+                for (int i = 0; i < instructions.Length; i++)
+                {
+                    string functionName = instructions[i].Symbol;
+                    if (funcToAddressRange.ContainsKey(functionName))
+                    {
+                        // Update the end address with this current instruction.
+                        funcToAddressRange[functionName].UpdateEndAddress(instructions[i]);
+                    }
+                    else
+                    {
+                        // Insert new function to keep track of.
+                        funcToAddressRange.Add(functionName, new DisasmAddressRange(instructions[i]));
+                    }
+
+                    // endAddr will not show up in instructions, so make the last instruction we get to get the range til the end.
+                    if (i == instructions.Length - 1)
+                    {
+                        funcToAddressRange[functionName].EndAddress = endAddr;
+                    }
+                }
+
+                foreach (string functionName in funcToAddressRange.Keys)
+                {
+                    DisasmAddressRange range = funcToAddressRange[functionName];
+
+                    // Mode 5 - mixed source and disassembly with raw opcodes
+                    cmd = "-data-disassemble -s " + EngineUtils.AsAddr(range.StartAddress, process.Is64BitArch) + " -e " + EngineUtils.AsAddr(range.EndAddress, process.Is64BitArch) + " -- 5";
+                    results = await process.CmdAsync(cmd, ResultClass.None);
+                    if (results.ResultClass != ResultClass.done)
+                    {
+                        return null;
+                    }
+                    try
+                    {
+                        /* Example response
+                         * asm_insns=[
+                         *  src_and_asm_line={
+                         *      line="15",
+                         *      file="main.cpp",
+                         *      fullname="/home/cpp/main.cpp",
+                         *      line_asm_insn=[ ... ]
+                         *  }
+                         * ]
+                         */
+                        ResultListValue asm_insns = results.Find<ResultListValue>("asm_insns");
```

Comment:
Not sure if throwing exception here when `Find` fails, slows down this process a lot.

Reply:
Seems like it is worth fixing to me.